# DIABETES PREDICTION

To import the important python libraries

In [67]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import sklearn as sk

import warnings
warnings.filterwarnings('ignore')

To upload the diabetes prediction dataset from the naan mudhalvan resources

In [68]:
diabetes=pd.read_csv('/content/karthie.csv')
diabetes.fillna(0)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

to get the head and tail of the dataset

In [69]:
diabetes.head(4)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0

In [70]:
diabetes.tail(4)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0

To explore the dataset using python

In [71]:
diabetes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


To describe the dataset using python

In [72]:
diabetes.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

To find the null values in the dataset

In [73]:
diabetes.isna().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

There is no any null values present in the dataset

To find the duplicated values in the dataset

In [74]:
diabetes.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
763    False
764    False
765    False
766    False
767    False
Length: 768, dtype: bool

To remove the all duplicated values in the dataset

In [75]:
diabetes.drop_duplicates(keep='first',inplace=True)
diabetes[diabetes.duplicated()]

Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

There is no any duplicated values in the dataset

To identify the shape of the dataset

In [76]:
diabetes.shape

(768, 9)

The above dataset contains the 768 rows and 9 columns present in the dataset

To find the datatypes in the dataset

In [77]:
diabetes.dtypes

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

To diaplay the all the columns present in the dataset

In [78]:
diabetes.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

to find the seperate columns values present in the dataset

1. Pregnancies

Diabetes can cause problems during pregnancy for women and their developing babies. Poor control of diabetes during pregnancy increases the chances for birth defects and other problems for the baby. It can also cause serious complications for the woman. Proper health care before and during pregnancy can help prevent birth defects and other health problems.

In [79]:
diabetes.Pregnancies

0       6
1       1
2       8
3       1
4       0
       ..
763    10
764     2
765     5
766     1
767     1
Name: Pregnancies, Length: 768, dtype: int64

2.gulcose

A fasting blood sugar level from 100 to 125 mg/dL (5.6 to 6.9 mmol/L) is considered prediabetes. If it's 126 mg/dL (7 mmol/L) or higher on two separate tests, you have diabetes. Glucose tolerance test. For this test, you fast overnight.

In [80]:
diabetes.Glucose

0      148
1       85
2      183
3       89
4      137
      ... 
763    101
764    122
765    121
766    126
767     93
Name: Glucose, Length: 768, dtype: int64

3.Blood pressure

Blood pressure target is usually below 140/90mmHg for people with diabetes or below 150/90mmHg if you are aged 80 years or above. For some people with kidney disease the target may be below 130/80mmHg. But it is important to speak to your healthcare team about your individual target.

In [81]:
diabetes.BloodPressure

0      72
1      66
2      64
3      66
4      40
       ..
763    76
764    70
765    72
766    60
767    70
Name: BloodPressure, Length: 768, dtype: int64

4.skin thickness

Skin thickness (epidermal surface to dermal fat inter- face), which is primarily determined by collagen con- tent, is greater in insulin-dependent diabetes mellitus (IDDM) patients who have been diabetic for >10 yr (11,12). This possibly reflects increased collagen cross- linkage and reduced collagen turnover (2,3).

In [82]:
diabetes.SkinThickness

0      35
1      29
2       0
3      23
4      35
       ..
763    48
764    27
765    23
766     0
767    31
Name: SkinThickness, Length: 768, dtype: int64

5.Insulin

With type 1 diabetes, the body does not make any insulin and therefore insulin has to be injected regularly every day to stay alive. With type 2 diabetes, the body does not make enough insulin, or the insulin that is made does not work well. Insulin injections are sometimes needed to manage blood glucose levels.

In [83]:
diabetes.Insulin

0        0
1        0
2        0
3       94
4      168
      ... 
763    180
764      0
765    112
766      0
767      0
Name: Insulin, Length: 768, dtype: int64

6.BMI

BMI-Body Mass Index

BMI is one of the most commonly used tools to assess whether you are overweight. So, if your BMI is at 30 or more, you are at risk of getting type 2 diabetes. The higher your BMI goes from this value, the greater the chances of getting type 2 diabetes.

In [84]:
diabetes.BMI

0      33.6
1      26.6
2      23.3
3      28.1
4      43.1
       ... 
763    32.9
764    36.8
765    26.2
766    30.1
767    30.4
Name: BMI, Length: 768, dtype: float64

7.diabetes pedigree function

Diabetes pedigree function (DPF) calculates diabetes likelihood depending on the subject's age and his/her diabetic family history. Very little is known about the determinants of DPF for gestational diabetes mellitus (GDM) and normal women.

In [85]:
diabetes.DiabetesPedigreeFunction

0      0.627
1      0.351
2      0.672
3      0.167
4      2.288
       ...  
763    0.171
764    0.340
765    0.245
766    0.349
767    0.315
Name: DiabetesPedigreeFunction, Length: 768, dtype: float64

8.Age

Type 2 diabetes most often develops in people over age 45, but more and more children, teens, and young adults are also developing it.

In [86]:
diabetes.Age

0      50
1      31
2      32
3      21
4      33
       ..
763    63
764    27
765    30
766    47
767    23
Name: Age, Length: 768, dtype: int64

9.Outcomes

Over time, diabetes can damage blood vessels in the heart, eyes, kidneys and nerves. People with diabetes have a higher risk of health problems including heart attack, stroke and kidney failure.

In [87]:
diabetes.Outcome

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

Now the data set will ready to apply the machine learning algorithm

to get the test and train data in the dataset

In [95]:
x=diabetes.drop('Outcome',axis=True)
y=diabetes['Outcome']
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.25)

In [96]:
x_train

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
555            7      124             70             33      215  25.5   
412            1      143             84             23      310  42.4   
396            3       96             56             34      115  24.7   
269            2      146              0              0        0  27.5   
407            0      101             62              0        0  21.9   
..           ...      ...            ...            ...      ...   ...   
732            2      174             88             37      120  44.5   
221            2      158             90              0        0  31.6   
141            5      106             82             30        0  39.5   
598            1      173             74              0        0  36.8   
53             8      176             90             34      300  33.7   

     DiabetesPedigreeFunction  Age  
555                     0.161   37  
412                     1.076   22  
396                     0.944   39  
269                     0.240   28  
407                     0.336   25  
..                        ...  ...  
732                     0.646   24  
221                     0.805   66  
141                     0.286   38  
598                     0.088   38  
53                      0.467   58  

[576 rows x 8 columns]

In [97]:
y_train

555    0
412    0
396    0
269    1
407    0
      ..
732    1
221    1
141    0
598    1
53     1
Name: Outcome, Length: 576, dtype: int64

In [98]:
x_test

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
659            3       80             82             31       70  34.2   
657            1      120             80             48      200  38.9   
77             5       95             72             33        0  37.7   
585            1       93             56             11        0  22.5   
402            5      136             84             41       88  35.0   
..           ...      ...            ...            ...      ...   ...   
423            2      115             64             22        0  30.8   
203            2       99             70             16       44  20.4   
316            3       99             80             11       64  19.3   
641            4      128             70              0        0  34.3   
522            6      114              0              0        0   0.0   

     DiabetesPedigreeFunction  Age  
659                     1.292   27  
657                     1.162   41  
77                      0.370   27  
585                     0.417   22  
402                     0.286   35  
..                        ...  ...  
423                     0.421   21  
203                     0.235   27  
316                     0.284   30  
641                     0.303   24  
522                     0.189   26  

[192 rows x 8 columns]

In [99]:
y_test

659    1
657    0
77     0
585    0
402    1
      ..
423    0
203    0
316    0
641    0
522    0
Name: Outcome, Length: 192, dtype: int64

to implementing the decision tree algorithm for the diabetes prediction for getting more accuracy than other algorithms

In [100]:
from sklearn.tree import DecisionTreeClassifier

In [101]:
dtree=DecisionTreeClassifier()
dtree=dtree.fit(x_train,y_train)

To predict the model

In [103]:
y_pred=dtree.predict(x_test)
y_pred

array([0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1,
       1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0,
       0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0])

To implement the confusion matrics and accuracy score

In [105]:
from sklearn.metrics import(confusion_matrix,classification_report,accuracy_score,ConfusionMatrixDisplay)

In [107]:
tree_accuracy=accuracy_score(y_test,y_pred)
tree_confusion=confusion_matrix(y_test,y_pred)
tree_display=ConfusionMatrixDisplay(confusion_matrix=tree_confusion)

In [108]:
tree_confusion

array([[98, 26],
       [36, 32]])

In [109]:
tree_display

Accuracy of the diabetes prediciton model

In [112]:
tree_accuracy

67.70833333333334

For diabetes prediction model the decision tree algorithm has accuracy of 68% of successful efficiency

Diabetes prediction using python excuted successfully